[receiver/sqlquery] add basic functionality

pmcollins opened the pull request:
This is the second PR for the new sqlquery receiver, containing the receiver's functionality, but without enabling the component.

First PR is here: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/9407

pmcollins committed:
[receiver/sqlqueryreceiver] add receiver functionality

rmfitzpatrick commented on receiver/cloudfoundryreceiver/config_test.go:
```diff
@@ -80,7 +80,7 @@ func TestLoadInvalidConfig(t *testing.T) {
 	factory := NewFactory()
 	factories.Receivers[typeStr] = factory
-	_, err = servicetest.LoadConfigAndValidate(filepath.Join("testdata", "config-invalid.yaml"), factories)
+	_, err = servicetest.LoadConfigAndValidate(filepath.Join("testdata", "config-invalid-datatype.yaml"), factories)
```
is this a stale update?

pmcollins replied:
not sure -- what makes you think it's stale?

pmcollins replied:
Oh I see. I changed that filename and GoLand probably updated that string. Will take a look (definitely don't want to be updating the cloud foundry receiver 😄).  Good catch.

pmcollins replied:
Should be reverted now.

rmfitzpatrick commented on receiver/sqlqueryreceiver/README.md:
````diff
@@ -34,15 +39,47 @@ receivers:
             attribute_columns: [ "genre" ]
 ```
+
+Given a `movie` table with three rows:
````
Nice example

rmfitzpatrick commented on receiver/sqlqueryreceiver/config.go:
```diff
@@ -28,16 +30,121 @@ type Config struct {
+		return errors.New("'query.metrics' cannot be empty")
+	}
+	for _, metric := range q.Metrics {
+		if err := metric.Validate(); err != nil {
```
Would aggregating in a multierr be helpful to report more issues at once?

pmcollins replied:
Good idea -- will fix.

pmcollins replied:
Done.

rmfitzpatrick commented on receiver/sqlqueryreceiver/db_client.go:
```diff
@@ -0,0 +1,112 @@
+	}
+	for _, columnType := range types {
+		colName := columnType.Name()
+		switch columnType.ScanType().Name() {
```
Mind linking to https://github.com/golang/go/blob/29057b707d468bd690e6eadfebe89c4ff737867c/src/database/sql/sql.go#L3210 in a comment?

Don't think it's necessary in these changes and I'm unsure of their prevalence but we could handle most of the uint cases. I wonder if floats should go in now.

crobert-1 commented on receiver/sqlqueryreceiver/receiver_test.go:
```diff
@@ -0,0 +1,56 @@
+	"go.uber.org/zap"
+)
+
+func TestCreateReceiver(t *testing.T) {
```
Would it be useful to test failure scenarios as well? It looks like the only time we'd get an error would be from `scraperhelper.NewScraperControllerReceiver` so I'm not sure how important it is to test that.

pmcollins replied:
Yeah, I think I'm missing test coverage for errors. Will add.

pmcollins replied:
I've increased test coverage for error conditions (though not in receiver.go since scraper controller handles errors and that's tested separately).

pmcollins committed:
[receiver/sqlqueryreceiver] generalize row scan retrieval logic to handle floats

pmcollins commented on receiver/sqlqueryreceiver/config.go:
```diff
@@ -28,16 +30,121 @@ type Config struct {
+		return errors.New("'metric.value_column' cannot be empty")
+	}
+	if err := c.ValueType.Validate(); err != nil {
+		return err
```
Good idea, will look into it.

pmcollins replied:
Done.

pmcollins commented on receiver/sqlqueryreceiver/config_test.go:
```diff
@@ -47,4 +47,34 @@ func TestParseConfig(t *testing.T) {
+			path.Join("testdata", cfgFile),
+			factories,
+		)
+		fmt.Printf("err: ->%v<-\n", err)
```
Nah, this was a straggler. Fixed.

pmcollins commented on receiver/sqlqueryreceiver/metrics.go:
```diff
@@ -0,0 +1,88 @@
+	dest.SetName(cfg.MetricName)
+	dest.SetDescription(cfg.Description)
+	dest.SetUnit(cfg.Unit)
+	dps := setMetricFields(cfg, dest)
+	dp := dps.AppendEmpty()
```
Sure, will do.

pmcollins replied:
Done.

pmcollins commented on receiver/sqlqueryreceiver/integration_test.go:
```diff
@@ -0,0 +1,116 @@
+	countByGenre := map[string]int64{}
+	extractGenre(t, ms.At(0), countByGenre)
+	extractGenre(t, ms.At(1), countByGenre)
+	assert.Equal(t, map[string]int64{"Action": 2, "SciFi": 3}, countByGenre)
```
Good question. `require.Eventually` blocks, and since it's a `require`, any statements after it wont run until and unless it succeeds, otherwise the test will fail and the statements after it won't run at all.

pmcollins committed:
[receiver/sqlqueryreceiver] improve test coverage; add multierr support for bad configs

pmcollins commented on receiver/sqlqueryreceiver/integration_test.go:
```diff
@@ -0,0 +1,116 @@
+	factory := NewFactory()
+	config := factory.CreateDefaultConfig().(*Config)
+	config.Driver = "postgres"
+	config.DataSource = "host=localhost port=" + externalPort + " user=otel password=otel sslmode=disable"
```
Done.

rmfitzpatrick commented on receiver/sqlqueryreceiver/config.go:
```diff
@@ -28,16 +31,126 @@ type Config struct {
+	Description      string            `mapstructure:"description"`
+}
+
+func (c MetricCfg) Validate() error {
```
I like this sub-config validation pattern

rmfitzpatrick commented on receiver/sqlqueryreceiver/config_test.go:
```diff
@@ -47,4 +44,48 @@ func TestParseConfig(t *testing.T) {
+			path.Join("testdata", cfgFile),
+			testFactories(t),
+		)
+		require.Error(t, err)
```
Would it be worth it to use a test table w/ expected error validation?

rmfitzpatrick commented on receiver/sqlqueryreceiver/scraper.go:
```diff
@@ -0,0 +1,78 @@
+	for _, metricCfg := range s.query.Metrics {
+		for _, row := range rows {
+			if err = rowToMetric(row, metricCfg, ms.AppendEmpty()); err != nil {
+				return out, fmt.Errorf("scraper: %w", err)
```
Is it better to error out immediately or try as is possible until the end?

pmcollins replied:
I suppose it'd give users more info if we combine all the errors. Fixed.

pmcollins committed:
[receiver/sqlqueryreceiver] use test table to check for errors; use multierr on Scrape

crobert-1 reviewed (approved):
Looks good, thanks for making all of the updates! ~~(Just need to update dependencies to fix the failing check)~~

pmcollins commented:
The failing checks are not related to this change. Anything more to do for this PR?

crobert-1 commented on receiver/sqlqueryreceiver/scraper.go:
```diff
@@ -0,0 +1,83 @@
+	for _, metricCfg := range s.query.Metrics {
+		for _, row := range rows {
+			if err = rowToMetric(row, metricCfg, ms.AppendEmpty()); err != nil {
+				err = multierr.Append(errs, err)
+			}
+		}
+	}
+	if err != nil {
+		errs = fmt.Errorf("scraper.Scrape errors: %w", errs)
+	}
```
Sorry for the late comment here, but I ran into an issue with this while doing some of my own testing. Currently we're merging `errs` and `err` into the `err` variable on line 71, repeatedly, for all errors we encounter. However, we're re-setting `err` every loop, meaning it'll only hold one error.

This means once we get to the check `if err != nil` we'll get into the if statement with just the last error. Then, in the `fmt.Errorf` statement we're using the `errs` variable, which hasn't actually been set at this point.

The current error message:
```
2022-06-23T15:14:23.606-0700	error	scraperhelper/scrapercontroller.go:197	Error scraping metrics	{"kind": "receiver", "name": "sqlquery", "error": "scraper.Scrape errors: %!w(<nil>)", "scraper": "sqlqueryreceiver/query-0: select count(*) as count, genre from sys.movie group by genre"}
```

I believe the solution here is to do `errs = multierr.Append(errs, err)` and then check `if errs != nil` below.

pmcollins replied:
oops, will fix.

pmcollins replied:
Done.

(I hadn't actually tested the contents of the errors so missed this dumb mistake. Added a test and made existing tests more explicit as well.)

pmcollins committed:
sqlqueryreceiver: fix multierr handling and make error tests explicit

crobert-1 commented on receiver/sqlqueryreceiver/config.go:
```diff
@@ -28,16 +31,126 @@ type Config struct {
+const (
+	MetricValueTypeUnspecified MetricValueType = ""
+	MetricValueTypeInt         MetricValueType = "int"
+	MetricValueTypeDouble      MetricValueType = "double"
```
Sorry, another late comment. The `README` says `float`, but this says `double`. I don't necessarily have a preference either way, it should just be consistent.

pmcollins replied:
Good catch, fixed. Going with float.

pmcollins committed:
sqlqueryreceiver: change metric value type from float to double

crobert-1 reviewed (approved):
Thanks for making the extra updates!

Konig-Corey commented on receiver/sqlqueryreceiver/config.go:
```diff
@@ -28,16 +31,126 @@ type Config struct {
+	Aggregation      MetricAggregation `mapstructure:"aggregation"`
+	Unit             string            `mapstructure:"unit"`
+	Description      string            `mapstructure:"description"`
+}
```
Do we want to consider adding an optional tags map that allow users to set custom tags on these metrics? I suppose they could always use a processor to do it but would be an easy add and would be nice to have the option.

pmcollins replied:
Hi @Konig-Corey, I think this could be done in a separate PR if other receivers are doing it as well. Otherwise I like the idea of letting a processor do it 🙂.

Konig-Corey replied:
@pmcollins  I submitted a PR on your branch if we think we want to add it. I think allowing someone to set them explicitly on the receiver makes sense instead of having to intercept these metrics in a processor to add tags. An example use case could be to add a tag(s) for table name, db instance etc.

pmcollins replied:
Are other receivers doing this as well? If so, can you submit the PR against the main branch after this one gets merged?

Konig-Corey replied:
@pmcollins no problem. Anyway we can expedite this getting merged in? It seems this PR has been open for quite some time.

Konig-Corey replied:
@dmitryax appears to be appending approval from you.

Konig-Corey commented on receiver/sqlqueryreceiver/README.md:
```diff
@@ -18,13 +18,18 @@ e.g. _host=localhost port=5432 user=me password=s3cr3t sslmode=disable_
 * `attribute_columns`(optional): a list of column names in the returned dataset used to set attibutes on the datapoint.
-* `is_monotonic`(optional): a boolean value indicating whether the metric value is monotonically increasing. If it is, the receiver will emit a sum type for this metric.
+* `data_type` (optional): can be `gauge` or `sum`; defaults to `gauge`.
+* `value_type` (optional): can be `int` or `float`; defaults to `int`.
```
looks like this is just incorrect verbiage in the readme, code parses a float64

Konig-Corey replied:
Right. The readme said float, but its parses a float64. Or am I missing something. I think you can see the PR I created if youd like to review the changes I made. Also, thank you for promptly following up on this!

pmcollins replied:
Sure, done.

dmitryax commented:
@Konig-Corey I believe you can push commits right into `pmcollins:sqlreceiver` branch. This also should be rebased

dmitryax commented:
I'm not sure, I worked for me every time I tried.

pmcollins committed:
sqlqueryrecevier: address PR feedback

Use 'float' instead of 'double' in config.
Disallow explicit aggregation when data type is gauge.

pmcollins commented on receiver/sqlqueryreceiver/README.md:
```diff
@@ -18,13 +18,18 @@ e.g. _host=localhost port=5432 user=me password=s3cr3t sslmode=disable_
+* `data_type` (optional): can be `gauge` or `sum`; defaults to `gauge`.
+* `value_type` (optional): can be `int` or `float`; defaults to `int`.
+* `monotonic` (optional): boolean; whether a cumulative sum's value is monotonically increasing (i.e. never rolls over or resets); defaults to false.
+* `aggregation` (optional): can be `cumulative` or `delta`; defaults to `cumulative`.
```
Good idea, done.
